### PR TITLE
ENH: Auto-close h5py.File using context manager

### PR DESCRIPTION
**Pull request by larsoner**
Use context managers where easy to avoid a bunch of warnings at Python exit of the form:
```
ResourceWarning: unclosed file <_io.BufferedRandom name=57>
```

### CODE REVIEW

File: gen/pysnirf2.jinja [hunk 1]
```diff
@@ -197,109 +197,109 @@
...
-            {% if TYPES.VARLEN_STRING in CHILD.type %}
-                    dataset = _create_dataset_string_array(tmp, '{{ CHILD.name }}', self._{{ CHILD.name }})
-                result._add(name, _validate_string_array(dataset, ndims=[2]))
-            {% elif TYPES.INT_VALUE in CHILD.type %}
-                    dataset = _create_dataset_int_array(tmp, '{{ CHILD.name }}', self._{{ CHILD.name }})
-                result._add(name, _validate_int_array(dataset, ndims=[2]))
-            {% elif TYPES.FLOAT_VALUE in CHILD.type %}
-                    dataset = _create_dataset_float_array(tmp, '{{ CHILD.name }}', self._{{ CHILD.name }})
-                result._add(name, _validate_float_array(dataset, ndims=[2]))
+        with h5py.File(TemporaryFile(), 'w') as tmp:
```
Comment (larsoner) [inline gen/pysnirf2.jinja:206]: This line was changed, all lines that follow were just indented one level

File: gen/pysnirf2.jinja [hunk 2]
```diff
@@ -390,25 +390,24 @@ class Snirf(Group):
...
             path = args[0]
             if not path.endswith('.snirf'):
                 path += '.snirf'
-            new_file = h5py.File(path, 'w')
-            self._save(new_file)
-            self._cfg.logger.info('Saved Snirf file at %s to copy at %s', self.filename, path)
-            new_file.close()
+            with h5py.File(path, 'w') as new_file:
+                self._save(new_file)
+                self._cfg.logger.info('Saved Snirf file at %s to copy at %s', self.filename, path)
```
Comment (larsoner) [inline gen/pysnirf2.jinja:410]: These lines were also indented

**Comment by sstucker** [comment]
This is great. On my todo list. does `__del__` seem to reliably close the files on your end? I found it does not in IPython

**Comment by larsoner** [comment]
> does __del__ seem to reliably close the files on your end? I found it does not in IPython

I actually implemented a `h5py_File_autoclose` class that inherited from `h5py.File` that just had a `__del__` method that called `self.close` at first. I *think* it worked at least some of the time (?), but settled on this since it's cleaner.

**Comment by sstucker** [comment]
Would it be more robust, maybe at expense of speed, to just open and close the file every time it is read and written?

**Comment by larsoner** [comment]
Could be. What's done now with the context manager seems clean enough, though. With the current implementation it's clear that the method opens one file, writes to it a bit, then closes it. And no need to worry about opening and closing repeatedly.

**Comment by sstucker** [comment]
The paradigm for the whole library, though, is that creating a `Snirf` instance opens a file in "read-write" mode, and calling `close()` on it or letting it exit scope should close the underlying `h5py.File` interface and calling `save()` without new arguments writes to the open file.

Calling `save(<with a new path>)` does as you describe
> opens one file, writes to it a bit, then closes it.

**Comment by sstucker** [comment]
> I actually implemented a h5py_File_autoclose class that inherited from h5py.File that just had a __del__ method that called self.close at first. I think it worked at least some of the time (?), but settled on this since it's cleaner.

Apparently `__del__` is not called when all references are gone or when an object is `del`ed but at the nondeterministic whims of the gc. So that explains the mixed results I was getting in IPython

**Pull request merged**